bundles/automation/org.eclipse.smarthome.automation.providers/src/main/java/org/eclipse/smarthome/automation/core/provider/RuleResourceBundleImporter.java

```diff
@@ -0,0 +1,229 @@
+     * sets {@code null} to {@link #ruleRegistry}.
+     * 
+     * @see AbstractResourceBundleProvider#close()
+     */
+    @Override
+    public void close() {
+        if (rulesTracker != null) {
+            rulesTracker.close();
+            rulesTracker = null;
+            ruleRegistry = null;
+        }
+        super.close();
+    }
+
+    /**
+     * @see AbstractResourceBundleProvider#addingService(ServiceReference)
+     */
+    @Override
+    public Object addingService(ServiceReference reference) {
+        if (reference.getProperty(Parser.PARSER_TYPE).equals(Parser.PARSER_RULE)) {
+            return super.addingService(reference);
+        }
+        return null;
+    }
+
+    /**
+     * This method provides functionality for processing the bundles with rule resources.
+     * <p>
+     * Checks for availability of the needed {@link Parser} and for availability of the rules managing service. If one
+     * of them is not available - the bundle is added into {@link #waitingProviders} and the execution of the method
+     * ends.
+     * <p>
+     * If both are available, the execution of the method continues with checking if the version of the bundle is
+     * changed. If the version is changed - removes persistence of old variants of the rules, provided by this bundle.
+     * <p>
+     * Continues with loading the new version of these rules. If this bundle is added for the very first time, only
+     * loads the provided rules.
+     * <p>
+     * The loading can fail because of {@link IOException}.
+     * 
+     * @param bundle it is a {@link Bundle} which has to be processed, because it provides resources for automation
+     *            rules.
+     */
+    @Override
+    protected void processAutomationProvider(Bundle bundle) {
+        String parserType = bundle.getHeaders().get(AutomationResourceBundlesEventQueue.AUTOMATION_RESOURCES_HEADER);
+        Parser parser = parsers.get(parserType);
+        if (parser == null || ruleRegistry == null) {
+            synchronized (waitingProviders) {
+                waitingProviders.put(new Long(bundle.getBundleId()), bundle);
+            }
+            return;
+        }
+        synchronized (providerPortfolio) {
+            for (Vendor vendor : providerPortfolio.keySet()) {
+                if (vendor.getVendorId().equals(Long.toString(bundle.getBundleId()))
+                        && !vendor.getVendorVersion().equals(bundle.getVersion().toString())) {
+                    remove(vendor.getVendorId());
+                    break;
+                }
+            }
+        }
+        Enumeration<URL> urls = bundle.findEntries(path, null, false);
+        if (urls == null) {
+            return;
+        }
+        ArrayList portfolio = new ArrayList();
+        while (urls.hasMoreElements()) {
+            URL url = urls.nextElement();
+            try {
+                Vendor vendor = new Vendor(Long.toString(bundle.getBundleId()), bundle.getVersion().toString());
```

Comment:
If a bundle contains several files in ESH-INF/automation/rules (as does the sample.handler.resources bundle), this vendor will be equal for each entry and thus you will get an IllegalArgumentException on the second entry:
```
Exception in thread "Automation Provider Processing Queue" java.lang.IllegalArgumentException: Cannot add element, because an element with same UID (87) already exists.
	at org.eclipse.smarthome.core.common.registry.AbstractManagedProvider.add(AbstractManagedProvider.java:56)
	at org.eclipse.smarthome.automation.core.provider.RuleResourceBundleImporter.importData(RuleResourceBundleImporter.java:223)
	at org.eclipse.smarthome.automation.core.provider.RuleResourceBundleImporter.processAutomationProvider(RuleResourceBundleImporter.java:162)
	at org.eclipse.smarthome.automation.core.provider.AutomationResourceBundlesEventQueue.processBundleChanged(AutomationResourceBundlesEventQueue.java:314)
	at org.eclipse.smarthome.automation.core.provider.AutomationResourceBundlesEventQueue.run(AutomationResourceBundlesEventQueue.java:138)
	at java.lang.Thread.run(Thread.java:722)
```